main/lsp/LSPTypechecker.cc

```diff
@@ -221,6 +221,126 @@ bool LSPTypechecker::typecheck(LSPFileUpdates updates, WorkerPool &workers,
+            continue;
+        }
+        if (fref.exists()) {
+            // Update to existing file on fast path
+            ENFORCE(fref.data(gs).getFileHash() != nullptr);
+            const auto &oldSymbolHashes = fref.data(gs).getFileHash()->localSymbolTableHashes;
+            const auto &newSymbolHashes = updatedFile->getFileHash()->localSymbolTableHashes;
+            const auto &oldMethodHashes = oldSymbolHashes.methodHashes;
+            const auto &newMethodHashes = newSymbolHashes.methodHashes;
+
+            if (config.opts.lspExperimentalFastPathEnabled) {
+                // Find which hashes changed. Note: methodHashes are sorted, so set_difference should work.
+                // This will insert two entries into `changedMethodHashes` for each changed method, but they
+                // will get deduped later.
+                absl::c_set_symmetric_difference(oldMethodHashes, newMethodHashes,
+                                                 std::back_inserter(changedMethodSymbolHashes));
+            } else {
+                // Both oldHash and newHash should have the same methods, since this is the fast path!
+                ENFORCE(validateIdenticalFingerprints(oldMethodHashes, newMethodHashes),
+                        "definitionHash should have failed");
+
+                // Find which hashes changed. Note: methodHashes are sorted, so set_difference should work.
+                // This will insert two entries into `changedMethodHashes` for each changed method, but they
+                // will get deduped later.
+                absl::c_set_difference(oldMethodHashes, newMethodHashes, std::back_inserter(changedMethodSymbolHashes));
+            }
+
+            const auto &oldFieldHashes = oldSymbolHashes.staticFieldHashes;
+            const auto &newFieldHashes = newSymbolHashes.staticFieldHashes;
+
+            ENFORCE(validateIdenticalFingerprints(oldFieldHashes, newFieldHashes), "definitionHash should have failed");
+
+            absl::c_set_difference(oldFieldHashes, newFieldHashes, std::back_inserter(changedFieldSymbolHashes));
+
+            result.changedFiles.emplace(fref, idx);
+        }
+    }
+
+    result.changedSymbolNameHashes.reserve(changedMethodSymbolHashes.size() + changedFieldSymbolHashes.size());
+    absl::c_transform(changedMethodSymbolHashes, std::back_inserter(result.changedSymbolNameHashes),
+                      [](const auto &symhash) { return symhash.nameHash; });
+    absl::c_transform(changedFieldSymbolHashes, std::back_inserter(result.changedSymbolNameHashes),
+                      [](const auto &symhash) { return symhash.nameHash; });
+    core::ShortNameHash::sortAndDedupe(result.changedSymbolNameHashes);
+
+    if (!result.changedSymbolNameHashes.empty()) {
+        // ^ optimization--skip the loop over every file in the project (`gs.getFiles()`) if
+        // the set of changed symbols is empty (e.g., running a completion request inside a
+        // method body)
```

Comment:
WDYT about dedenting this now that the logic is in its own function?

```c++
if (result.changedSymbolNameHashes.empty()) {
  return result;
}

// ...logic to further modify `result`
return result;
```

Reply:
I ended up doing this in the stacked PR that follows.